keypad/kpc_agent.py

```diff
@@ -1,33 +1,106 @@
 """File contains the KPC agent"""
+from functools import reduce
+
+from finite_state_machine.rule import signal_is_digit
+from keypad.keypad import Keypad
+from keypad.led_board import LEDBoard
 
 
 class KPCAgent:
     """The agent communicating with the finite state machine, keypad and led board"""
 
+    def __init__(self):
+        self.led_board = LEDBoard()
+        self.led_board.setup()
+        self.keypad = Keypad()
+        self.override_signal = ""
+        self._current_password = []
+        self._password = ""
+        self._temp_password = []
+
+        self._led_id = None
+        self._led_dur = None
+
     def init_passcode_entry(self):
         """ Clear the passcode-buffer and initiate a ”power up” lighting sequence on the LED
         Board"""
+        self._current_password = []
+        self.led_board.start_up_sequence()
+        self.override_signal = ""
 
     def get_next_signal(self) -> str:
         """ Return the override-signal, if it is non-blank; otherwise query the keypad for the next
         pressed key"""
+        symbol = self.keypad.get_next_signal()
+        if self.override_signal != "":
+            return self.override_signal
+        return symbol
 
-    def _verify_login(self):
+    def append_password(self, char: str):
+        """ Append a char to the current password """
+        self._current_password.append(char)
+
+    def verify_login(self):
         """Check that the password just entered via the keypad matches that in the password file.
         Store the result (Y or N) in the override-signal and light the leds appropriately"""
+        current_password_as_string = convert_list_to_string(self._current_password)
+        self.override_signal = (
+            "Y" if current_password_as_string == self._password else "N"
+        )
+        # TODO add signal to show user whether login was correct or not
+
+    def compare_new_passwords(self):
+        """ Check that the new passwords match """
+        if self._current_password == self._temp_password:
+            self.update_passcode_file()
+            # TODO add signal to show user if passwords were equal
+
+    def cache_first_password(self):
+        """ Cache first password in order to compare passwords later """
+        self._temp_password = self._current_password
+        self._current_password = []
```

Comment:
I'm guessing these two are the password and confirm password values. if so, change the names to something a bit clearer